lizrd/text/packers.py

```diff
@@ -0,0 +1,206 @@
+        sample_end = sample_start + self.sequence_length
+
+        target_ids = list(take_circular(target_ids, sample_start, sample_end))
+        input_ids = list(take_circular(input_ids, sample_start, sample_end))
+        calculate_loss = list(take_circular(calculate_loss, sample_start, sample_end))
+
+        return LLMExample(input_ids, target_ids, calculate_loss)
+
+    def _mask_text(self, tokens: List[int]) -> Tuple[List[int], List[int]]:
+        mask_id = self.tokenizer.mask_id
+        assert mask_id is not None
+
+        is_mask = self.np_rng.binomial(
+            1, self.mask_replace_config.mask_percentage, len(tokens)
+        )
+        how_to_mask = self.np_rng.multinomial(
+            1,
+            [
+                self.mask_replace_config.replace_with_mask,
+                self.mask_replace_config.replace_with_random,
+                self.mask_replace_config.replace_with_original,
+            ],
+            size=len(tokens),
+        ).nonzero()[1]
+        replacements = (
+            (how_to_mask == 0) * mask_id
+            + (how_to_mask == 1) * self._get_valid_random_tokens(len(tokens))
+            + (how_to_mask == 2) * tokens
+        )
+
+        input_tokens = np.where(is_mask.astype(bool), replacements, tokens)
+
+        return input_tokens.tolist(), is_mask.tolist()
+
+    def _get_valid_random_tokens(self, tokens_count):
+        assert isinstance(self.tokenizer, BertTokenizer)
+        special_tokens = 999
+        return (
+            self.np_rng.choice(self.tokenizer.vocab_size - special_tokens, tokens_count)
+            + special_tokens
+        )
+
+
+class GPTPacker(
+    AbstractPacker,
+):
+    def __init__(
+        self,
+        sequence_length: int,
+        dataset: AbstractDataset,
+        tokenizer_maker: Callable[[], AbstractTokenizer],
+        seed: Optional[int] = None,
+    ):
+        super().__init__(
+            sequence_length,
+            dataset,
+            tokenizer_maker,
+            seed=seed,
+        )
+
+    def get_sample(self) -> LLMExample:
+        """
+        Sample examples from the dataset until we reach the desired sequence length.
+        """
+        eot_id = self.tokenizer.eot_id
+        assert eot_id is not None
+
+        buffer: List[int] = []
+        calculate_loss: List[int] = []
+        document_lengths: List[int] = []
+
+        while True:
+            document = self.dataset.get_document()
+            tokens = self.tokenizer.text_to_ids(document)
+            buffer.extend(tokens + [eot_id])
+
+            document_lengths.append(len(tokens) + 1)
+            if (sum(document_lengths) - max(document_lengths)) > self.sequence_length:
+                break
+
+        sample_start = self.py_rng.randint(0, len(buffer) - 1)
+        sample_end = sample_start + self.sequence_length
+
+        input_ids = list(take_circular(buffer, sample_start, sample_end))
+        target_ids = list(take_circular(buffer, sample_start + 1, sample_end + 1))
```

Comment:
Nie rozumiem czemu chcemy wprost zwracać listę?

Reply:
LLMExample trzyma listę i imho to jest dobrze - to powinna być rzecz, w której możemy indeksować, już taki konkretny obiekt.